sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+        XX = row_norms(X_out_chunk, squared=True)[:, np.newaxis]
+        YY = X_in_norm_squared
+        dist_out_in = safe_sparse_dot(X_out_chunk, X_in.T, dense_output=True)
+        dist_out_in *= -2
+        dist_out_in += XX
+        dist_out_in += YY
+
+        ind1, = np.where((dist_out_in < margin_radii_out[chunk, None]).ravel())
+        ind2, = np.where((dist_out_in < margin_radii_in[None, :]).ravel())
+        ind = np.unique(np.concatenate((ind1, ind2)))
+
+        if len(ind):
+            ind_plus_offset = ind + chunk.start * len(X_in)
+            try:
+                imp_ind.extend(ind_plus_offset)
+            except TypeError:
+                imp_ind.append(ind_plus_offset)
+
+            if return_distance:
+                # This np.maximum would add another ~8% time of computation
+                np.maximum(dist_out_in, 0, out=dist_out_in)
+                dist_chunk = dist_out_in.ravel()[ind]
+                try:
+                    dist.extend(dist_chunk)
+                except TypeError:
+                    dist.append(dist_chunk)
+
+    imp_ind = np.asarray(imp_ind)
+
+    if return_distance:
+        return imp_ind, np.asarray(dist)
+    else:
+        return imp_ind
+
+
+def paired_distances_batch(X, ind_a, ind_b, mem_budget=int(1e7)):
+    """Equivalent to  row_norms(X[ind_a] - X[ind_b], True)
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        An array of data samples.
+    ind_a : array, shape (n_indices,)
+        An array of sample indices.
+    ind_b : array, shape (n_indices,)
+        Another array of sample indices.
+    mem_budget : int, optional (default=int(1e7))
+        Memory budget (in bytes) for computing distances.
+    Returns
+    -------
+    distances: array, shape (n_indices,)
+        An array of pairwise squared distances.
+    """
+
+    bytes_per_row = X.shape[1] * X.itemsize
+    batch_size = int(mem_budget // bytes_per_row)
+
+    n_pairs = len(ind_a)
+    distances = np.zeros(n_pairs)
+    for chunk in gen_batches(n_pairs, batch_size):
+        distances[chunk] = row_norms(X[ind_a[chunk]] - X[ind_b[chunk]], True)
+
+    return distances
+
+
+def sum_outer_products(X, weights):
+    """Computes the sum of weighted outer products using a sparse weights
+    matrix
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features)
+        An array of data samples.
+
+    weights : csr_matrix, shape (n_samples, n_samples)
+        A sparse weights matrix (indicating target neighbors).
+
+
+    Returns
+    -------
+    sum_outer_prods : array, shape (n_features, n_features)
+        The sum of all weighted outer products.
+    """
+
+    weights_sym = weights + weights.T
+    diag = spdiags(weights_sym.sum(1).ravel(), 0, *weights_sym.shape)
+    laplacian = diag - weights_sym
+    sum_outer_prods = X.T.dot(laplacian.dot(X))
+
+    return sum_outer_prods
+
+
+def check_scalar(x, name, dtype, min_val=None, max_val=None):
```

Comment:
private?

I understand the motivation but having such check functions for such specific estimator seems unfortunate to me.